Elements/src/Spatial/AdaptiveGrid/AdaptiveGrid.cs

```diff
@@ -1330,14 +1330,110 @@ private void SplitGrid(Grid2d grid, IEnumerable<Vector3> keyPoints)
             grid.V.SplitAtPoints(keyPoints);
         }
 
-        private HashSet<Edge> AddFromGrid(Grid2d grid, IEnumerable<Edge> edgesToIntersect)
+        private List<double> gridUDividers(Grid1d grid)
         {
-            var cells = grid.GetCells();
+            if (grid == null)
+            {
+                return new List<double>();
+            }
+            if (grid.IsSingleCell)
+            {
+                return new List<double> { grid.Domain.Min, grid.Domain.Max };
+            }
+            var ans = grid.Cells.Select(gridUDividers).Select(l => l.First()).ToList();
+            ans.Add(grid.Cells.Last().Domain.Max);
+            return ans;
+        }
+
+        private Transform gridFromUVTransform(Grid2d grid)
+        {
+            var uDomain = grid.U.curveDomain;
+            var vDomain = grid.V.curveDomain;
+            Vector3 xAxis = (grid.U.Evaluate(uDomain.Max) - grid.U.Evaluate(uDomain.Min)) / uDomain.Length;
+            Vector3 yAxis = (grid.V.Evaluate(vDomain.Max) - grid.V.Evaluate(vDomain.Min)) / vDomain.Length;
+            Vector3 zAxis = xAxis.Cross(yAxis);
+            var fromGrid = new Transform(grid.U.Evaluate(uDomain.Min) + grid.V.Evaluate(vDomain.Min), xAxis, yAxis, zAxis).Concatenated(grid.fromGrid);
+            return fromGrid;
+        }
+
+        private List<(Vector3 from, Vector3 to)> splitSegmentWithPoints((Vector3 from, Vector3 to) line, List<Vector3> points)
+        {
+            var resultingSegments = new List<(Vector3 from, Vector3 to)>();
+            var lineVector = line.to - line.from;
+            points.Add(line.from);
+            points.Add(line.to);
+            points.Sort((p, q) => Math.Sign(lineVector.Dot(p - q)));
+            for (int i = 1; i < points.Count; ++i)
+            {
+                resultingSegments.Add((points[i - 1], points[i]));
+            }
+            return resultingSegments;
+        }
+
+        private List<(Vector3 from, Vector3 to)> splitSegmentsWithPoints(List<Line> segments, double u, List<double> coords, bool coordsAreX, List<Vector3> intersectionPoints)
+        {
+            var swapXYAxes = new Transform(new Vector3(0, 0, 0), Vector3.YAxis, Vector3.XAxis, Vector3.ZAxis);
+            if (coordsAreX)
+            {
+                segments = segments.Select(l => l.TransformedLine(swapXYAxes)).ToList();
```

Comment:
Again, we have 3 to lists applied to segments when we can have none:
using ienumerable variable, for example "sortedSegmets" you can transform it when reverse when order it by Y coordinate and then foreach it without need of extra memory allocation.